docs/user_config.md

```diff
@@ -0,0 +1,150 @@
+layout: docs
+---
+
+A Tiltfile can define a user config, which allow the users of a Tiltfile to provide
+input / configuration to that Tiltfile. For example, one might write a Tiltfile
+such that one runs `tilt up consumer` to run Tilt with the services needed for
+development for consumers, or `tilt up enterprise` to run Tilt with the services
+needed for development for enterprise.
+
+This doc describes how to describe your setup to Tilt to enable these
+shorthands. We aim to make common cases trivial, while also supporting escape
+hatches that enable arbitrarily-complex cases.
+
+In order, we'll cover:
+* Examples that cover common cases, showing both the Tiltfiles and accompanying
+  usage.
+* Describe how this is implemented. These details, especially around
+  persistence, will let you understand how to write tooling that interacts with
+  this system.
+* Future work (we see how it fits in but don't plan to implement until there's
+  concrete demand).
+
+### Run only some services
+Your app has many services, call them A, B, C, and D, and you want your users to
+be able to run only a subset of them. E.g., `tilt up` will run all services, but
+running `tilt up a b d` will run only the resources a, b, and d, ignoring c.
+This is the same as the default `tilt up` behavior.
+
+#### Command-Lines
+* `tilt up`: run all services
+* `tilt up a b d`: run A, B, and D but not C
+* `tilt flags a b`: change running Tilt to run A and B but neither C nor D.
+* `tilt up`: run all services
+
+#### Tiltfile
+```python
+config.define_string_list("to-run", args=True)
+cfg = config.parse()
+config.set_enabled_resources(cfg.get('to-run', []))
+```
+
+### Run a defined set of services
+Your app has many services, call them A, B, C, and D. You have defined modes you
+want users to be able to start (e.g. consumer vs. enterprise) and you don't want
+them to have to remember which service is necessary for which mode.
+You want to be able to write a doc that tells users to run `tilt up consumer`
+the first time and know that's enough that they'll be set up. Running Tilt
+subsequent times without arguments should default to the mode they already have
+set.
+
+#### Command-Lines
+* `tilt up`: run all services
+* `tilt up consumer`: run only services in the "consumer" list
+#### Tiltfile
+```python
+config.define_string_list("to-run", args=True)
+cfg = config.parse()
+groups = {
+  'consumer': ['a', 'b', 'c'],
+  'enterprise': ['a', 'b', 'd'],
+}
+resources = []
+for arg in cfg.get('args', []):
+  if arg in groups:
+    resources += groups[arg]
+  else:
+    # also support specifying individual services instead of groups, e.g. `tilt up a b d`
+    resources.append(arg)
+config.set_enabled_resources(resources)
+```
+
+### Specify services to edit
+You have many services that your users frequently need to run, but don't expect
+to edit. Your YAML includes good enough images. (Perhaps they're from a suitably
+recent CI invocation, or refer to standard images.)
```

Comment:
> Your YAML includes good enough images. (Perhaps they're from a suitably recent CI invocation, or refer to standard images.)

try:

> Your YAML references pre-existing images (perhaps they're from a suitably
recent CI invocation, or refer to standard images) such that the user can run the app without doing a Docker Build via Tilt

Reply:
Maybe add something like:

> This pattern allows you to run the majority of your services from pre-existing images (avoiding the overhead of a Docker Build), while using Tilt to watch files and quickly update services you're actively developing on.